// cohere/dist/relationship.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var ensureValidArgs = function ensureValidArgs(name, inverse, options) {
  if (typeof name !== 'string') {
    throw new TypeError('Argument "name" must be a string.');
  }

  if (typeof inverse !== 'string') {
    throw new TypeError('Argument "inverse" must be a string.');
  }

  if ((typeof options === 'undefined' ? 'undefined' : _typeof(options)) !== 'object') {
    throw new TypeError('Argument "options" must be an object.');
  }
};

var relationship = exports.relationship = function relationship(relation) {
  return function (name, inverse) {
    var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

    if (!['belongsTo', 'hasOne', 'hasMany'].includes(relation)) {
      throw new Error('Argument "relation" must be either "belongsTo", "hasOne", or "hasMany", instead got ' + ('"' + relation + '"'));
    }

    ensureValidArgs(name, inverse, options);

    return function (field) {
      return _extends({}, options, {
        inverse: { name: null, relation: null, field: inverse },
        relation: relation,
        name: name,
        field: field
      });
    };
  };
};

var hasMany = exports.hasMany = relationship('hasMany');
var belongsTo = exports.belongsTo = relationship('belongsTo');
var hasOne = exports.hasOne = relationship('hasOne');
