// cohere/dist/Schema.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _Type = require('./Type');

var _Type2 = _interopRequireDefault(_Type);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Schema = function () {
  function Schema() {
    var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';

    _classCallCheck(this, Schema);

    this.isCompiled = false;
    this._types = {};

    this.name = name;
  }

  _createClass(Schema, [{
    key: 'defineType',
    value: function defineType(name, fields) {
      if (Object.keys(this.types).includes(name)) {
        throw new Error('The "' + name + '" type has already been defined.');
      }

      this._types[name] = new _Type2.default(name, fields);
      return this;
    }
  }, {
    key: 'compile',
    value: function compile() {
      var _this = this;

      // hydrate every inverse
      this.types.forEach(function (type) {
        var name = type.name;


        type.relationships.forEach(function (relationship) {
          var inverse = relationship.inverse,
              relation = relationship.relation,
              field = relationship.field,
              relatedName = relationship.name;


          if (!_this.types[relatedName]) {
            throw new Error('Could not compile the schema, because the "' + name + '" type\'s "' + field + '" ' + ('relationship has a name "' + relatedName + '", which has not been defined. Try running ') + ('schema.defineType(' + relatedName + ', { ... }).'));
          }

          _this.types[relatedName].hydrateInverse(inverse.field, {
            name: relatedName,
            relation: relation
          });
        });
      });

      // link types together
      this.types.forEach(function (type) {
        type.relationships.forEach(function (relationship) {
          relationship.type = _this.types[relationship.name]; // eslint-disable-line
        });
      });

      this.isCompiled = true;
      return this;
    }
  }, {
    key: 'types',
    get: function get() {
      var types = this._types;

      return _extends({}, types, {
        forEach: function forEach(fn) {
          Object.keys(types).forEach(function (name) {
            return fn(types[name]);
          });
        },
        reduce: function reduce(fn, initialValue) {
          return Object.keys(types).reduce(function (init, name) {
            return fn(init, types[name]);
          }, initialValue);
        }
      }, ['map', 'some', 'every', 'filter', 'find'].reduce(function (accumulator, iterator) {
        return _extends({}, accumulator, _defineProperty({}, iterator, function (fn) {
          return Object.keys(types)[iterator](function (name) {
            for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
              args[_key - 1] = arguments[_key];
            }

            return fn.apply(undefined, [types[name]].concat(args));
          });
        }));
      }, {}));
    }
  }]);

  return Schema;
}();

exports.default = Schema;
