// cohere/dist/Type.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var ensureUniqueFields = function ensureUniqueFields() {
  var attributes = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var relationships = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  var attributesFields = Object.keys(attributes);
  var relationshipsFields = Object.keys(relationships);

  if (attributesFields.some(function (field) {
    return relationshipsFields.includes(field);
  })) {
    throw new Error('Attribute field names and relationship field names must be unique.');
  }
};

var Type = function () {
  function Type(name) {
    var _this = this;

    var fields = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, Type);

    if (typeof name !== 'string') {
      throw new TypeError('Argument "name" must be a valid string.');
    }

    if (!(fields && (typeof fields === 'undefined' ? 'undefined' : _typeof(fields)) === 'object')) {
      throw new TypeError('Argument "fields" must be a valid object.');
    }

    ensureUniqueFields(fields.attributes, fields.relationships);

    this.name = name;
    this._attributes = fields.attributes || {};
    this._relationships = fields.relationships || {};

    // hydrate relationships
    this._relationships = Object.keys(this._relationships).reduce(function (prev, field) {
      var relationship = _this._relationships[field];

      if (typeof relationship !== 'function') {
        throw new TypeError('Tried registering the "' + name + '" type\'s "' + field + '" relationship, but it wasn\'t a ' + 'function. Please use either the "hasMany", "belongsTo", or "hasOne" relationship ' + 'function.');
      }

      return _extends({}, prev, _defineProperty({}, field, relationship(field)));
    }, {});

    // hydrate all other top level keys
    Object.keys(fields).forEach(function (key) {
      if (key !== 'attributes' && key !== 'relationships') {
        _this[key] = fields[key];
      }
    });
  }

  _createClass(Type, [{
    key: 'hasAttribute',
    value: function hasAttribute(field) {
      return this._attributes.hasOwnProperty(field);
    }
  }, {
    key: 'hasRelationship',
    value: function hasRelationship(field) {
      return this._relationships.hasOwnProperty(field);
    }
  }, {
    key: 'attribute',
    value: function attribute(field) {
      return this.attributes[field] || null;
    }
  }, {
    key: 'relationship',
    value: function relationship(field) {
      return this.relationships[field] || null;
    }
  }, {
    key: 'hydrateInverse',
    value: function hydrateInverse(field, _ref) {
      var name = _ref.name,
          relation = _ref.relation;

      if (!this.relationship(field)) {
        throw new Error('Tried to hydrate the "' + this.name + '" type\'s "' + field + '" inverse, but "' + field + '" is not a ' + 'valid relationship.');
      }

      var relationship = this.relationship(field);

      if (!relationship.inverse.isHydrated) {
        this._relationships[field].inverse = _extends({}, relationship.inverse, {
          isHydrated: true,
          name: name,
          relation: relation
        });
      }
    }
  }, {
    key: 'attributes',
    get: function get() {
      var attributes = this._attributes;

      return _extends({}, attributes, {
        forEach: function forEach(fn) {
          Object.keys(attributes).forEach(function (field) {
            return fn({
              type: attributes[field],
              field: field
            });
          });
        },
        reduce: function reduce(fn, initialValue) {
          return Object.keys(attributes).reduce(function (init, field) {
            return fn(init, {
              type: attributes[field],
              field: field
            });
          }, initialValue);
        }
      }, ['map', 'some', 'every', 'filter', 'find'].reduce(function (accumulator, iterator) {
        return _extends({}, accumulator, _defineProperty({}, iterator, function (fn) {
          return Object.keys(attributes)[iterator](function (field) {
            for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
              args[_key - 1] = arguments[_key];
            }

            return fn.apply(undefined, [_extends({}, attributes[field], { field: field })].concat(args));
          });
        }));
      }, {}));
    }
  }, {
    key: 'relationships',
    get: function get() {
      var relationships = this._relationships;

      return _extends({}, relationships, {
        forEach: function forEach(fn) {
          Object.keys(relationships).forEach(function (field) {
            return fn(relationships[field]);
          });
        },
        reduce: function reduce(fn, initialValue) {
          return Object.keys(relationships).reduce(function (init, field) {
            return fn(init, relationships[field]);
          }, initialValue);
        }
      }, ['map', 'some', 'every', 'filter', 'find'].reduce(function (accumulator, iterator) {
        return _extends({}, accumulator, _defineProperty({}, iterator, function (fn) {
          return Object.keys(relationships)[iterator](function (field) {
            for (var _len2 = arguments.length, args = Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
              args[_key2 - 1] = arguments[_key2];
            }

            return fn.apply(undefined, [relationships[field]].concat(args));
          });
        }));
      }, {}));
    }
  }]);

  return Type;
}();

exports.default = Type;
